pdfannots/printer/json.py

```diff
@@ -2,79 +2,36 @@
+class JsonPrinter(DictBasedPrinter):
     def __init__(self, args: argparse.Namespace):
         super().__init__(args)
         self.remove_hyphens = args.remove_hyphens  # Whether to remove hyphens across a line break
 
-        # We overload the printfilename option to decide whether to emit a structured
-        # JSON file with one entry per file, or just a flat array of annotations.
-        self.printfilename = args.printfilename
+        # Multiple input files in a single output require to have a filename
+        self.printfilename = args.printfilename if (
+            not (hasattr(args, "input") and len(args.input) > 1)) else True
         self.seen_first = False
 
     def begin(self) -> str:
-        return '[\n' if self.printfilename else ''
+        return '{\n' if self.printfilename else ''
 
     def end(self) -> str:
-        return '\n]\n' if self.printfilename else '\n'
-
-    def print_file(
-        self,
-        filename: str,
-        document: Document
-    ) -> typing.Iterator[str]:
+        return '\n}\n' if self.printfilename else '\n'
 
+    def print_file(self, filename: str, document: Document) -> typing.Iterator[str]:
         # insert a , between successive files
         if self.seen_first:
-            if self.printfilename:
-                yield ',\n'
-            else:
-                # The flat array format is incompatible with multiple input files
-                # TODO: Ideally we'd catch this at invocation time
-                raise RuntimeError("When used with multiple input files, the "
-                                   "JSON formatter requires --print-filename")
+            yield ',\n'
         else:
             self.seen_first = True
 
         annot_dicts = [self.annot_to_dict(document, a) for a in document.iter_annots()]
 
         if self.printfilename:
-            dictobj = {
-                "file": filename,
-                "annotations": annot_dicts
-            }
-
-        yield from json.JSONEncoder(indent=2).iterencode(
-            dictobj if self.printfilename else annot_dicts)
-
-    def annot_to_dict(self, doc: Document, annot: Annotation) -> typing.Dict[str, typing.Any]:
-        """Convert an annotation to a dictionary representation suitable for JSON encoding."""
-        assert annot.pos
-
-        result = {
-            "type": annot.subtype.name,
-            "page": annot.pos.page.pageno + 1,
-            "start_xy": (annot.pos.x, annot.pos.y),
-        }
-
-        outline = doc.nearest_outline(annot.pos)
-        if outline:
-            result["prior_outline"] = outline.title
-
-        if annot.text:
-            result['text'] = annot.gettext(self.remove_hyphens)
-
-        if annot.contents:
-            result['contents'] = annot.contents
-
-        if annot.author:
-            result['author'] = annot.author
-
-        if annot.created:
-            result['created'] = annot.created.strftime('%Y-%m-%dT%H:%M:%S')
+            yield '  "%s": ' % filename
```

Comment:
Ok, I see what you've done. This is reasonable if we do keep the multi-file support, but I don't think it's necessary.